### ISSUE TITLE
Docs: Overhaul no-floating-promise messages and docs

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#8088]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

Right now, [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises) logs quite a mouthful for its messages:

```plaintext
Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator.
```

So much text! Longer messages are less likely to be read by users - especially "error" (or "warning") messages like this one.

In addition, it's also missing the point that storing the Promise in a variable is considered an acceptable workaround by the rule (assuming you have unused variable detection enabled). https://github.com/typescript-eslint/typescript-eslint/discussions/8046 is an example of a discussion asking about that.

This is a nuanced area. See replies to: https://twitter.com/flybayer/status/1469098848958222337, https://twitter.com/threepointone/status/1471458830626299911, https://twitter.com/nandafyi/status/1736164705448956388, and other threads from searching on `no-floating-promises`). I think we should rework the rule a bit to coach/encourage developers into treating the nuance with care:

* Make the error message much shorter (so folks will actually read the whole thing), and perhaps encourage them to read the docs for more info
* Mention the unused variable allowance in the docs
* Link to a blog post explaining how to enable our Promise/Thenable-area rules (perhaps a separate issue?)

### Affected URL(s)

https://typescript-eslint.io/rules/no-floating-promises

**Comment by JoshuaKGoldberg** [comment]
Since #7008 was added to the 8 milestone for tracking, adding this as well. I think we'll really need to have good docs to make sure folks don't misuse this.

**Comment by JoshuaKGoldberg** [comment]
Btw, referencing https://github.com/typescript-eslint/typescript-eslint/issues/7008 & https://github.com/nodejs/node/issues/51292 & https://github.com/reduxjs/redux-toolkit/pull/4102: we should definitely mention strategies for frameworks/libraries to mark Promises as safe (or not expose them).

**Comment by JoshuaKGoldberg** [comment]
Note: #8573 has a good suggestion:

> I fixed a no-floating-promises error by changing:
> 
> ```
> - promise.catch(() => {}).then(() => {})
> + promise.catch(() => {}).finally(() => {})
> ```
> 
> I _think_ this is a good change! Since my `.catch()` handler couldn't reject, it's probably not even a change.
> 
> The `.catch().then()` form is [mentioned on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#chaining_after_a_catch) and triggers this rule ([playground](https://typescript-eslint.io/play#ts=5.3.3&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBKRBvAUIxNAQwGcBPSQrAXgD4d8DEZhCBZIqACwDoAnImAAmcALaZEAHkQAGHgFYseJk2584Ad0RgAploCifdX0wBuRgQC%2BjSxkwXEPCJwhc0OxHQYqCEBCTgAGx0eHSM4Ex0McxVbHm4dMDQJL2UfPzAA4J5AuABzNAByIkDAxDgAa0Loh1tzayA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)). So it would be helpful if the docs on this rule included an example of the suggested refactor.

**Comment by JoshuaKGoldberg** [comment]
Marking as blocked on PRs implementing #7008 and #8404, as I'd neglected to note this publicly till now. 🙂

**Comment by JoshuaKGoldberg** [comment]
Now that #9186 is merged, I believe this is unblocked. I'd like to tackle it alongside #8517.

#8404 would be great too.

**Comment by benmccann** [comment]
> When Not To Use It
> if you're not worried about crashes from floating or misused Promises

It would be nice for there to be some discussion of browser vs server in the docs. I believe that at least in the past that Node would crash upon there being an unhandled promise rejection though I don't know if that's still the case. Do we need to worry about it in the browser or is this only a Node.js thing? I see on StackOverflow that [browsers have not swallowed such errors in a long time](https://stackoverflow.com/questions/52061394/do-browsers-still-swallow-unhandled-rejected-promises-silently-what-about-node).

I also wonder about browser/server with [`no-misused-promises`](https://typescript-eslint.io/rules/no-misused-promises/). A large fraction of the times it triggers is on browser event handlers:
```
	addEventListener('popstate', async (event) => {
		...
	});
```

Does this matter? The docs say the rule triggers when "they are not handled properly". But what would be the proper handling here? I wonder if this rule is almost always a false positive in browser code like this or if there's something I'm actually supposed to be doing here.

**Comment by JoshuaKGoldberg** [comment]
> in the past that Node would crash upon there being an unhandled promise rejection

Throwing on unhandled rejections by default was shipped by Node.js in its v15: https://github.com/nodejs/node/issues/20392 -> https://github.com/nodejs/node/commit/3b10f7f933dee3607649b6bd1ca24876704222ac -> https://github.com/nodejs/node/releases/tag/v15.0.0.

> browser/server

We should note that whether the runtime (browser, edge worker, full server, ...) handles the Promise isn't the main point of these rules. Roughly all runtimes provide some way of being _notified_ of unhandled rejections (e.g. [the `unhandledrejection` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event), but handling them _in application logic_ is a separate matter. The idea that some area of code might _start_ some work and _not handle that work's rejection_ matters.

> Does this matter?

Great question, and something we'll definitely want to address in the docs. Yes it does! If that `async (event) => { ... }` function rejects, nothing will handle that rejection by default. `addEventListener` doesn't know to.

```ts
declare function doSomeWork(): Promise<void>;

addEventListener("popstate", async (event) => {
  await doSomeWork();
  throw new Error("Who should handle me?");
});
```

> But what would be the proper handling here?

Another great question we should handle in the docs! There's no one-size-fits-all solution because how each application might handle generalized or specific errors is different. Two common approaches we've seen have been:

- Having the equivalent of a root-level `try catch` and/or `.finally`
- Including a function like the following that registers errors in some app-specific handler:
  ```ts
  function handlePromiseSafely(doWork: () => Promise<unknown>): void {
    doWork().catch((error: unknown) => {
      // (or however you'd handle this)
      console.error("Oh no:", error);
    });
  }
  ```

**Comment by benmccann** [comment]
Thanks so much for the detailed and thoughtful response!

>Yes it does! If that async (event) => { ... } function rejects, nothing will handle that rejection by default. addEventListener doesn't know to.

Isn't that basically the same with a synchronous handler?
```
declare function doSomeWork(): void;

addEventListener("popstate", (event) => {
  doSomeWork();
  throw new Error("Who should handle me?");
});
```

When this rule is used to bubble up exceptions to `try`/`catch` blocks it feels useful. But when it's bubbling up exception to event handlers it just feels it's turning all promises into checked exceptions, which can be a bit annoying, and I don't know why async functions should be treated differently than synchronous functions in that case.

> Including a function like the following that registers errors in some app-specific handler

Don't we already have this with [`onunhandledrejection`](https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event)? Why not just use that instead of wrapping every individual promise instance?

**Comment by JoshuaKGoldberg** [comment]
> Isn't that basically the same with a synchronous handler?

For the case specifically of the DOM APIs, sure, yes. But if the user has their own:

```ts
function runAction(action: () => void) {
  console.log("before");
  action();
  console.log("this should only happen if the action didn't crash");
}

declare function doSomeWork: () => Promise<void>;

runAction(async () => {
  await doSomeWork();
  throw new Error("Who should handle me?");
});
```

> Why not just use that instead of wrapping every individual promise instance?

That's one strategy that works for handling floating/miscellaneous exceptions. But see above for control flow issues.

https://jakearchibald.com/2023/unhandled-rejections has more funky situations.

---

Anyway, @benmccann you raise reasonable things to discuss but this getting off-topic for the specific issue of overhauling the docs. If you want to chat more could I ask you to make a Discussion or thread in the Discord? We can dive into deeper points there.

**Comment by JoshuaKGoldberg** [comment]
Per https://github.com/typescript-eslint/typescript-eslint/issues/8517#issuecomment-2237031622, moving this to post-v8.
